gh-119609, PEP 756: Add PyUnicode_Export() function (#123738)
python/cpython: Doc/c-api/unicode.rst

   .. versionadded:: 3.3

> .. c:function:: int PyUnicode_Export(PyObject *unicode, int32_t requested_formats, Py_buffer *view)

Reviewer: Update also the return type?

   .. versionadded:: 3.3

.. c:function:: int PyUnicode_Export(PyObject *unicode, uint32_t requested_formats, Py_buffer *view, uint32_t *format)

>    Export the contents of the *unicode* string in one of the requested format
>    *requested_formats*.

Reviewer 2: ```suggestion
   Export the contents of the *unicode* string in one of the *requested_formats*.
```

>    * On success, fill *view* and set *\*format*, and return ``0``.
>    * On error, set an exception and return ``-1``.

Reviewer 2: ```suggestion
   * On success, fill *view*, set *\*format*, and return ``0``.
   * On error, set an exception, set *\*format* to 0, and return ``-1``.
     *view* is left unchanged.
```

>    The *view* buffer must be released by :c:func:`PyBuffer_Release`.

Reviewer 2: ```suggestion
   After a successful call to :c:func:`PyUnicode_Export`,
   the *view* buffer must be released by :c:func:`PyBuffer_Release`.
```

   The contents of the buffer are valid until they are released.

   The buffer is read-only and must not be modified.

   *unicode*, *view* and *format* must not be NULL.

   Available formats:

   .. c:namespace:: NULL

   ===================================  ========  ===========================
   Constant Identifier                  Value     Description
   ===================================  ========  ===========================
   .. c:macro:: PyUnicode_FORMAT_ASCII  ``0x01``  ASCII string (``Py_UCS1*``)
   .. c:macro:: PyUnicode_FORMAT_UCS1   ``0x02``  UCS-1 string (``Py_UCS1*``)
   .. c:macro:: PyUnicode_FORMAT_UCS2   ``0x04``  UCS-2 string (``Py_UCS2*``)
   .. c:macro:: PyUnicode_FORMAT_UCS4   ``0x08``  UCS-4 string (``Py_UCS4*``)
   .. c:macro:: PyUnicode_FORMAT_UTF8   ``0x10``  UTF-8 string (``char*``)

>    ===================================  ========  ===========================

Reviewer 2: ```suggestion
   ===================================  ========  ===========================
   
   UCS-2 and UCS-4 use the native byte order.
```

   .. versionadded:: 3.3

.. c:function:: int PyUnicode_Export(PyObject *unicode, uint32_t requested_formats, Py_buffer *view, uint32_t *format)

   Export the contents of the *unicode* string in one of the *requested_formats*.

   * On success, fill *view*, set *\*format*, and return ``0``.
   * On error, set an exception, set *\*format* to 0, and return ``-1``.
     *view* is left unchanged.

   After a successful call to :c:func:`PyUnicode_Export`,
   the *view* buffer must be released by :c:func:`PyBuffer_Release`.
   The contents of the buffer are valid until they are released.

   The buffer is read-only and must not be modified.

   *unicode*, *view* and *format* must not be NULL.

   Available formats:

   .. c:namespace:: NULL

   ===================================  ========  ===========================
   Constant Identifier                  Value     Description
   ===================================  ========  ===========================

>    .. c:macro:: PyUnicode_FORMAT_ASCII  ``0x01``  ASCII string (``Py_UCS1*``)
>    .. c:macro:: PyUnicode_FORMAT_UCS1   ``0x02``  UCS-1 string (``Py_UCS1*``)
>    .. c:macro:: PyUnicode_FORMAT_UCS2   ``0x04``  UCS-2 string (``Py_UCS2*``)
>    .. c:macro:: PyUnicode_FORMAT_UCS4   ``0x08``  UCS-4 string (``Py_UCS4*``)
>    .. c:macro:: PyUnicode_FORMAT_UTF8   ``0x10``  UTF-8 string (``char*``)

Reviewer 2: ```suggestion
   .. c:macro:: PyUnicode_FORMAT_ASCII  ``0x10``  ASCII string (``Py_UCS1*``)
   .. c:macro:: PyUnicode_FORMAT_UCS1   ``0x01``  UCS-1 string (``Py_UCS1*``)
   .. c:macro:: PyUnicode_FORMAT_UCS2   ``0x02``  UCS-2 string (``Py_UCS2*``)
   .. c:macro:: PyUnicode_FORMAT_UCS4   ``0x04``  UCS-4 string (``Py_UCS4*``)
   .. c:macro:: PyUnicode_FORMAT_UTF8   ``0x08``  UTF-8 string (``char*``)
```

Author: Oops, updated.

   .. versionadded:: 3.3

.. c:function:: int PyUnicode_Export(PyObject *unicode, uint32_t requested_formats, Py_buffer *view)

   Export the contents of the *unicode* string in one of the requested format
   *requested_formats*.

   * On success, fill *view*, and return ``0``.
   * On error, set an exception and return ``-1``.

>    The export must be released by :c:func:`PyBuffer_Release`.

Reviewer 3: ```suggestion
   The *view* must be released by :c:func:`PyBuffer_Release`.
```

   The contents of the buffer are valid until they are released.

   The buffer is read-only and must not be modified.

   *unicode* and *view* must not be NULL.

   Available formats:

   .. c:namespace:: NULL

   ===================================  ========  ===========================
   Constant Identifier                  Value     Description
   ===================================  ========  ===========================
   .. c:macro:: PyUnicode_FORMAT_ASCII  ``0x01``  ASCII string (``Py_UCS1*``)
   .. c:macro:: PyUnicode_FORMAT_UCS1   ``0x02``  UCS-1 string (``Py_UCS1*``)
   .. c:macro:: PyUnicode_FORMAT_UCS2   ``0x04``  UCS-2 string (``Py_UCS2*``)
   .. c:macro:: PyUnicode_FORMAT_UCS4   ``0x08``  UCS-4 string (``Py_UCS4*``)
   .. c:macro:: PyUnicode_FORMAT_UTF8   ``0x10``  UTF-8 string (``char*``)
   ===================================  ========  ===========================

>    *requested_formats* can be a single format or a bitwise combination of the
>    formats in the table above.
>    On success, *\*format* will be set to a single one of the requested flags.

Reviewer 3: ```suggestion
   *requested_formats* can be a single format or a bitwise combination of the
   formats in the table above.
   On success, *\*format* will be set to a single one of the requested flags.
```

I'm not sure I understand. What is the *format*? When you say "one of the requested", is there any order of precedence? for instance, if you want to export it to `PyUnicode_FORMAT_UCS1 | PyUnicode_FORMAT_UCS2 | PyUnicode_FORMAT_UCS4` and your unicode string is actually a USC2, does it mean that the final format would be `PyUnicode_FORMAT_UCS2`?

>    On success, *\*format* will be set to a single one of the requested flags.

Reviewer 4: Here, `*format` is one of the arguments to `PyUnicode_GetBufferFormat`, right?  So maybe:

To determine the format that was selected for output, call `PyUnicode_GetBufferFormat`.

   Note that future versions of Python may introduce additional formats.

   .. versionadded:: 3.14

.. c:function:: int PyUnicode_GetBufferFormat(const Py_buffer *view, uint32_t *format)

   Get the format of the buffer *view*.

>    * On success, set *\*result* to the corresponding `PyUnicode_FORMAT_*` value

Reviewer 4: ```suggestion
   * On success, set *\*format* to the corresponding `PyUnicode_FORMAT_*` value
```

     and return ``0``.
   * On error, set an exception and return ``-1``.

   *view* must be a buffer filled by :c:func:`PyUnicode_Export`.

   .. versionadded:: 3.14

.. c:function:: PyObject* PyUnicode_Import(const void *data, Py_ssize_t nbytes, uint32_t format)

>    Create a string object from a buffer in an “export format”.

Reviewer 4: ```suggestion
   Create a *unicode* string object from a buffer in a supported format.
```
